raft/src/main/java/org/apache/kafka/raft/FuturePurgatory.java

```diff
@@ -20,37 +20,44 @@
 
 /**
  * Simple purgatory interface which allows tracking a set of expirable futures.
+ * Each future will be associated with a comparable value which are used to determine
+ * if the future object can be completed.
+ *
+ * Note that the future objects should be organized in order so that completing awaiting
+ * futures would stop early and not traverse all the futures
  *
  * @param <T> Type completion type
  */
-public interface FuturePurgatory<T> {
+public interface FuturePurgatory<T extends Comparable<T>> {
 
     /**
      * Add a future to this purgatory for tracking.
      *
      * @param future the future tracking the expected completion. A subsequent call
-     *               to {@link #completeAll(Object)} will complete this future
+     *               to {@link #completeAll(Comparable, long)} will complete this future
      *               if it does not expire first.
+     * @param value  the comparable value of the future object that will be used to determine
+     *               if the future can be completed or not
      * @param maxWaitTimeMs the maximum time to wait for completion. If this
      *               timeout is reached, then the future will be completed exceptionally
      *               with a {@link org.apache.kafka.common.errors.TimeoutException}
      */
-    void await(CompletableFuture<T> future, long maxWaitTimeMs);
+    void await(CompletableFuture<Long> future, T value, long maxWaitTimeMs);
 
     /**
      * Complete all awaiting futures. The completion callbacks will be triggered
      * from the calling thread.
      *
-     * @param value the value that will be passed to {@link CompletableFuture#complete(Object)}
-     *              when the futures are completed
+     * @param value       the threshold value used to determine which futures can be completed
+     * @param currentTime the current time in milli seconds that will be passed to {@link CompletableFuture#complete(Object)}
+     *                    when the futures are completed
      */
-    void completeAll(T value);
+    void completeAll(T value, long currentTime);
```

Comment:
We should probably change this name so that it is clear that only futures exceeding the threshold value will be completed. Maybe we can just drop `All` from the name?